### PR TITLE
Add git to run requirements

### PR DESCRIPTION
**Pull request by epruesse**
`import git` will actually fail if no `git` binary can be located.

* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy`
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by epruesse** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -20,6 +20,7 @@ requirements:
   run:
     - python
     - gitdb2
+    - git
```
Comment (ocefpaf) [inline recipe/meta.yaml:23]: I believe that the decision was to follow what the PyPI does and expect the user to install `git` separately. This avoids wasting space for those who already have `git` via their system while it is still for others to just `conda install git` if they do not have it.

Comment (epruesse) [reply recipe/meta.yaml:23]: I added it because it breaks in Bioconda in the mulled tests, often quite unexpectedly deep down the chain when something makes something else import gitpython and then an exception is raised.

Is there a way to have this marked in big bold letters as a caveat of the package to be aware of?

Comment (ocefpaf) [reply recipe/meta.yaml:23]: The package does take care of that my issuing a "git not found" error. Not sure what else we could do.

Comment (epruesse) [reply recipe/meta.yaml:23]: Thinking about this a bit more. You wrote:
>I believe that the decision was to follow what the PyPI does and expect the user to install git separately. 

Actually, the PyPI packages doesn't because it can't. Git isn't Python and PyPI packages are limited to Python dependencies. We, however, can. Conda fixes exactly that shortcoming of pip. So we can do as e.g. Debian's `python-git` package does, and install `git` where needed.

What exactly is the benefit of having `git` from outside the environment? Why would someone want to do that? And how is git different from any other tool that we package and automatically install if it is required as dependency?

Git isn't particularly huge. It doesn't have a massive dependency chain. It is also not something that would be installed by default on (non-developer) boxes. So I would strongly argue in favor of adding git to gitpython as a dependency. That way, it will just work. The conda way. :grin:

Comment (jakirkham) [reply recipe/meta.yaml:23]: Seems like a reasonable argument to me.

I wasn't able to find the discussion Filipe is referring to, but wouldn't be surprised if there was one. That context might be useful for this discussion as well if someone can point to it.

Comment (ocefpaf) [reply recipe/meta.yaml:23]: It was during one of our early meetings. I wanted to add `git` and you and Phil were against it b/c most people already have `git` the package is built to handle an external installation of it. I don't think it is "the conda way" to add all "optional" packages but I do agree that sometimes it makes sense. Like when we add `pandoc` for `nbconvert` on `jupyter`, but `pandoc` is not something that is commonly available on the user system like `git`.

TL;DR is that I don't really care but I do believe that people who are trying to make our packages lighter and more cloud friendly will.

Comment (epruesse) [reply recipe/meta.yaml:23]: > I don't think it is "the conda way" to add all "optional" packages

No distro does. In this case, `gitpython` is unusable without the `git` binary installed. It raises an exception at `import git`. So it's not an optional dependency. 

It's possible to alleviate the cloud image size issue by reducing the size of the `git` package. Looking at the OSX package, the binaries are 37MB unpacked at the moment. 6.7MB of that are copies (!) of `git`, `git-cvsserver`, `git-shell` and `git-upload-pack` kept in both `bin` and `libexec/git-core`. Using symlinks and running `strip` on `libexec/git-core/*` drops the size to 27MB. Curiously, the `git` binaries are statically linked against an internal `libgit.a`. I'm pretty certain that using a DSO instead would drop total size down a lot further. I really don't know why they do it this way - `git` is a tough word to search things on the actual tool.

Comment (epruesse) [reply recipe/meta.yaml:23]: I just hacked the makefile to link the git binaries against a `libgit.dylib`. Now the whole thing is 4.5MB unpacked.

Comment (jakirkham) [reply recipe/meta.yaml:23]: Thanks for the info, Filipe.

Agree that people targeting cloud deployment or even embedded systems will care about the size from adding this. TBH conda-forge also cares about this in the context of our webservices, which live on Heroku and are constantly at risk of going over tight size constraints.

Elmar I think Filipe is right that we can consider git to some extent "optional" (borrowing his quotes here). It's not optional in the traditional sense of the package not *needing* it, but it is in the sense of many systems already providing a copy and some packages (like on PyPI) not shipping it.

Agree with using split packages to address package size. We have used this to some extent already. It will probably be useful for other situations as well. We just need to figure out a way to do this that doesn't massively increase the maintenance burden of each package.

Probably worth noting Windows is a complication here. We may want to choose `m2-git` instead of `git`. Worth getting some Windows users perspectives on this.

Anyways I don't have strong feelings here either. It's an easy enough issue for people to work around. Though don't object to adding it if it helps.

Comment (epruesse) [reply recipe/meta.yaml:23]: The reason I came here was a Bioconda package using snakemake which in turn uses gitpython which then fails in the "clear" testing container that comes without anything. The question is then should the package using snakemake, or snakemake, or snakemake-minimal, or gitpython depend on git. Given that it will not even import, my feeling was that it's gitpython.

I don't know about the windows package. I could have a go at git for osx/linux, trying to slim it down. 4.5M unpacked should be fine in most instances. There are bigger space hogs to fry (like not having `-dbgsyms` packages / not stripping everything). My brief testing did yield a slower startup time for git. About a 20% increase, so 0.002 seconds. I don't think there are many applications in which this matters. I would have to ask on the git mailing list about the choice to link each binary statically, though.

Comment (jakirkham) [reply recipe/meta.yaml:23]: On Windows we just repackage an existing PortableGit binary in the `git` package (it precedes the MSYS2 packages). It works, but I'm guessing it is the wrong solution here. Would just use `m2-git` for Windows. Though if we can get some Windows users to come along and tell us what they want that would be better.

Comment (bgruening) [reply recipe/meta.yaml:23]: Mh, so we have @ocefpaf and @jakirkham that are +-0 on this and we have @epruesse and me with a +1. Should we put this issue for other comments in gitter or are we fine in adding it and see if someone really complains about the theoretical space issue?

Comment (epruesse) [reply recipe/meta.yaml:23]: I've started a [discussion at git-scm](https://public-inbox.org/git/13c40b4b819f702a52f7039177579f87fa90aa50.camel@mad-scientist.net/t/#u), asking about using dynamic linking to make git itself smaller. It appears that from a technical standpoint, there is no reason to not do that. With the package being only 5MB installed, the main argument against disappears.

Comment (jakirkham) [reply recipe/meta.yaml:23]: The other thing to keep in mind is `gitpython` is a dependency of `conda-smithy`. So `conda-smithy` users that have their `base` environment active would wind up having their system `git` shadowed by ours.

**Comment by yarikoptic** [comment]
FWIW +1 from me to add git as runtime dependency since it is required and provided via conda.  pypi argument is indeed weak since it doesn't provide git.

**Comment by yarikoptic** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml (hunk 1 above)
Comment (igortg) [reply recipe/meta.yaml:23]: Hey guys, I was just lurking  arround here and saw this thread.

I'm not aware of any company/user that install Git through Conda. Normally, a distro that already includes Git is used, or a provision step that install Git is made before anything else. I'm not saying using conda for that is wrong, but I believe that  the most common way is to have it already installed.

So having a package that installs the conda git package is a no-go for a big part of conda/conda-forge users.

I would suggest to decline this PR

**Comment by xhochy** [comment]
Closing this as the discussion in https://github.com/conda-forge/gitpython-feedstock/pull/25#discussion_r292161489 yielded quite a lot of opponents to it. If there are hard feeling, we should raise with `conda-forge/core`.
